fendermint/vm/topdown/src/sync.rs

```diff
@@ -0,0 +1,205 @@
+    pub fn start(self) -> anyhow::Result<()> {
+        let config = self.config;
+        let provider = self.parent_view_provider;
+        let agent = self.agent;
+
+        tokio::spawn(async move {
+            loop {
+                if let Err(e) = sync_with_parent(&config, &agent, &provider).await {
+                    tracing::info!("sync with parent encountered error: {e}");
+                }
+            }
+        });
+
+        Ok(())
+    }
+}
+
+/// Extract the actual error from Box
+macro_rules! downcast_err {
+    ($r:expr) => {
+        match $r {
+            Ok(v) => Ok(v),
+            Err(e) => match e.downcast_ref::<Error>() {
+                None => unreachable!(),
+                Some(e) => Err(e.clone()),
+            },
+        }
+    };
+}
+
+async fn sync_with_parent<T: ParentFinalityProvider + Send + Sync + 'static>(
+    config: &Config,
+    agent_proxy: &Arc<IPCAgentProxy>,
+    provider: &Arc<T>,
+) -> anyhow::Result<()> {
+    let mut interval = tokio::time::interval(Duration::from_secs(config.polling_interval_secs));
+
+    loop {
+        interval.tick().await;
+
+        let (starting_height, latest_height) =
+            get_query_block_heights(agent_proxy, provider).await?;
+        if latest_height < config.chain_head_delay {
+            tracing::debug!("latest height not more than the chain head delay");
+            continue;
+        }
+
+        // we are going backwards in terms of block height, the latest block height is lower
+        // than our previously fetched head. It could be a chain reorg. We clear all the cache
+        // in `provider` and start from scratch
+        if starting_height > latest_height {
+            todo!()
+        }
+
+        let new_parent_views =
+            get_new_parent_views(agent_proxy, starting_height, latest_height).await?;
+        let r = atomically_or_err(move || {
+            for (height, block_hash, validator_set, messages) in new_parent_views.clone() {
+                provider.new_parent_view(height, block_hash, validator_set, messages)?;
+            }
+            Ok(())
+        })
+        .await;
+        downcast_err!(r)?;
+    }
+}
+
+/// Obtain the block height range to perform the parent view update
+async fn get_query_block_heights<T: ParentFinalityProvider + Send + Sync + 'static>(
+    agent_proxy: &Arc<IPCAgentProxy>,
+    provider: &Arc<T>,
+) -> anyhow::Result<(BlockHeight, BlockHeight)> {
+    // update block hash and validator set
+    let latest_height = agent_proxy
+        .get_chain_head_height()
+        .await
+        .context("cannot fetch parent chain head")?;
+
+    let r = atomically_or_err(|| {
+        Ok(if let Some(h) = provider.latest_height()? {
+            h + 1
+        } else {
+            let last_committed_finality = provider.last_committed_finality()?;
+            last_committed_finality.height + 1
+        })
+    })
+    .await;
+    let starting_height = downcast_err!(r)?;
```

Comment:
How is the starting height retrieved? IIRC from the previous PR the `ParentFinalityProvider` doesn't fallback on-chain to see what was the latest height committed. If this is the case, what happens if the node is syncing from scratch a the finality for height `n >> 0` is already committed. Will this start fetching blocks from `0` to the latest committed?

I think we should maybe include the fallback here

Reply:
As discussed over slack, when it first start, it will fetch from the chain, but the storage on chain has yet to be implemented, so I left the part on fetching from storage unimplemented.